articles/cross_compilation_build_tools_design.md

```diff
@@ -0,0 +1,248 @@
+  by ROS 2, or building on a host with different platforms than the target
+  platform. That later use case is particularly useful for speeding up build
+  times by building the code on a powerful development workstation, and then
+  deploying the binaries to a constrained resources hardware that uses the
+  target platform, thus making the development cycle much faster.
+author: >
+  [Thomas Moulard](https://github.com/thomas-moulard),
+  [Juan Rodriguez Hortala](https://github.com/juanrh)
+published: false
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+The [cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation)
+contains instructions for cross-compiling ROS 2 to unsupported platforms, using
+some configuration files for CMake and Docker that is setting maintained in the
+[ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
+However, that tutorial doesn't provide instructions to build other ROS 2
+packages, but just for building a whole ROS 2 distribution. Ideally, we should
+be able to cross-compile a ROS package for ROS 2 by launching a single
+`colcon build` command. Finally, cross-compiling for other platforms besides
+ARM64 is not supported in the [ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
+
+## Proposed Approach
+
+We propose continuing with the general approach of the cross-compilation
+tutorial, based on building a sysroot for the target platform using qemu and
+Docker, and then using `cmake-toolchains` with a toolchain file pointing
+`CMAKE_SYSROOT` to the sysroot location to compile with C and C++
+cross-compilers.
+
+On top of that we propose using a **colcon plugin `cc-build`** to wrap the
+instructions on the cross-compilation tutorial into a single build command.
+That plugin would define a new `colcon_core.verb` entry point pointing to a
+class `CCBuildVerb` that uses the [Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern)
+to wrap the [`BuildVerb` class](https://github.com/colcon/colcon-core/blob/master/colcon_core/verb/build.py#L74)
+used by `colcon build` with additional functionality to handle the sysroot and
+toolchain file, and to inject additional `cmake-args` arguments. That would
+allows us to run the single command below from a ROS 2 overlay workspace on a
+developer workstation, for cross-compiling the packages in the workspace up to
+a package `performance_test` for the ARM-HF platform.
+
+```bash
+colcon cc-build --arch armhf --os ubuntu_bionic \
+  --packages-up-to performance_test # other `colcon build` arguments
+```
+
+That would should result on a new workspace directory `armhf-ubuntu_bionic`
+with subdirectories `build`, `install`, and `log` directories, with the
+ross-compiled binaries. All options for `colcon build` will be supported by
```

Comment:
ross => cross

I actually propose to proof read this PR for spelling mistakes and other grammar.

Reply:
I have made a pass to fix spelling and grammar mistake, please let me know if you find something else